File: app/src/main/java/com/mirash/familiar/activity/pin/view/PinIndicatorView.java
package com.mirash.familiar.activity.pin.view;

import android.content.Context;
import android.util.AttributeSet;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.mirash.familiar.Const;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Mirash
 */
public class PinIndicatorView extends LinearLayout {
    private List<PinCircleView> pinIndicatorViews;

    public PinIndicatorView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public PinIndicatorView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        setOrientation(HORIZONTAL);
        int size = Const.PIN_CODE_SIZE;
        setWeightSum(size);
        pinIndicatorViews = new ArrayList<>(size);
        for (int i = 0; i < size; i++) {
            PinCircleView pinCircleView = new PinCircleView(getContext());
            pinIndicatorViews.add(pinCircleView);
            addView(pinCircleView, new LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1));
        }
    }

    public void setCheckedSize(int checkedSize) {
        for (int i = 0; i < pinIndicatorViews.size(); i++) {
            pinIndicatorViews.get(i).setChecked(i < checkedSize);
        }
    }
}

File: app/src/main/java/com/mirash/familiar/activity/pin/view/PinIndicatorView.kt
package com.mirash.familiar.activity.pin.view

import android.content.Context
import android.util.AttributeSet
import android.view.ViewGroup
import android.widget.LinearLayout
import com.mirash.familiar.tool.PIN_CODE_SIZE

/**
 * @author Mirash
 */
class PinIndicatorView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {

    private val pinIndicatorViews: MutableList<PinCircleView>

    init {
        orientation = HORIZONTAL
        val size = PIN_CODE_SIZE
        weightSum = size.toFloat()
        pinIndicatorViews = ArrayList(size)
        for (i in 0 until size) {
            val pinCircleView = PinCircleView(context)
            pinIndicatorViews.add(pinCircleView)
            addView(pinCircleView, LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f))
        }
    }

    fun setCheckedSize(checkedSize: Int) {
        for (i in pinIndicatorViews.indices) {
            pinIndicatorViews[i].isChecked = i < checkedSize
        }
    }
}
